src/main/java/frc/robot/subsystems/SwerveWheelModuleSubsystem.java

```diff
@@ -0,0 +1,141 @@
+    private final double I = .00002;
+
+    private TalonFX angleMotor;
+    private TalonFX speedMotor;
+    private PIDController pidController;
+    private CANcoder angleEncoder;
+    // private boolean calibrateMode;
+    private double encoderOffset;
+    private String motorName;
+    // private SimpleWidget speedLim;
+
+    public SwerveWheelModuleSubsystem(int angleMotorChannel, int speedMotorChannel, int angleEncoderChannel,
+            String motorName, double offset) {
+        // We're using TalonFX motors on CAN.
+        this.angleMotor = new TalonFX(angleMotorChannel);
+        this.speedMotor = new TalonFX(speedMotorChannel);
+        this.angleEncoder = new CANcoder(angleEncoderChannel); // CANCoder Encoder
+        this.speedMotor.setNeutralMode(NeutralModeValue.Coast);
+        // angleMotor.configSupplyCurrentLimit(Constants.kdriveCurrentLimit);
+        // speedMotor.configSupplyCurrentLimit(Constants.kdriveCurrentLimit);
+        this.motorName = motorName;
+
+        this.pidController = new PIDController(P, I, 0); // This is the PID constant,
+        // we're not using any
+        // Integral/Derivative control but increasing the P value will make
+        // the motors more aggressive to changing to angles.
+
+        pidController.enableContinuousInput(0, 360); // This makes the PID controller
+        // understand the fact that for
+        // our setup, 360 degrees is the same as 0 since the wheel loops.
+
+        SendableRegistry.addChild(this, angleMotor);
+        SendableRegistry.addChild(this, speedMotor);
+        SendableRegistry.addChild(this, angleEncoder);
+        SendableRegistry.addLW(this, "Swerve Wheel Module");
+
+        encoderOffset = offset;
+        resetSensor();
+    }
+
+    public void drive(double speed, double angle) {
+        double currentEncoderValue = getPosition();
+        int reverse = setAngle(angle, currentEncoderValue);
+        setSpeed(speed * reverse);
+
+        SmartDashboard.putNumber("Swerve CANCoder " + motorName, getPosition());
+    }
+
+    public int setAngle(double angle, double currentEncoderValue) {
+        angle = MathUtil.mod(angle, 360); // ensure setpoint is on scale 0-360
+        int reverse = 1;
+        // angle += 90;
+
+        if (MathUtil.getCyclicalDistance(currentEncoderValue, angle, 360) > 70) {
+            reverse = -1;
+            angle += 180;
+            angle = MathUtil.mod(angle, 360);
+        }
+
+        double pidOut = -pidController.calculate(currentEncoderValue, angle);
+
+        angleMotor.set(pidOut);
+
+        return reverse;
+    }
+
+    public void setSpeed(double speed) {
+        speedMotor.set(speed); // sets motor speed //22150 units/100 ms at 12.4V
+    }
+
+    // this method outputs position of the encoder to the smartDashBoard, useful for
+    // calibrating the encoder offsets
+    public double getPosition() {
+        return MathUtil.mod(angleEncoder.getAbsolutePosition().getValue() - encoderOffset, 360);

```

Comment:
if we are having problems with wheels constantly spinning, lets verify this function is correctly returning the position of the wheels. the can encoder api changed.

Best to do this via the periodic method, just put the output of this func to the smart dashboard.